pkg/util/inplaceupdate/inplace_update_vertical.go

```diff
@@ -0,0 +1,252 @@
+	}
+
+	if state.LastContainerStatuses == nil {
+		state.LastContainerStatuses = make(map[string]appspub.InPlaceUpdateContainerStatus)
+	}
+	c := state.LastContainerStatuses[container.Name]
+	if state.LastContainerStatuses[container.Name].Resources.Limits == nil {
+		c.Resources.Limits = make(map[v1.ResourceName]resource.Quantity)
+	}
+	if state.LastContainerStatuses[container.Name].Resources.Requests == nil {
+		c.Resources.Requests = make(map[v1.ResourceName]resource.Quantity)
+	}
+
+	if container.Resources != nil {
+		for key, quantity := range container.Resources.Limits {
+			if !v.CanResourcesResizeInPlace(string(key)) {
+				continue
+			}
+			c.Resources.Limits[key] = quantity
+		}
+		for key, quantity := range container.Resources.Requests {
+			if !v.CanResourcesResizeInPlace(string(key)) {
+				continue
+			}
+			c.Resources.Requests[key] = quantity
+		}
+	}
+	// Store container infos whether c is empty or not.
+	// It will be used in IsContainerUpdateCompleted to check container resources is updated.
+	// case: pending pod can also be resize resource
+	state.LastContainerStatuses[container.Name] = c
+}
+
+// UpdateResource implements vertical updates by directly modifying the container's resources,
+// conforming to the k8s community standard
+func (v *VerticalUpdate) UpdateContainerResource(container *v1.Container, newResource *v1.ResourceRequirements) {
+	if container == nil || newResource == nil {
+		return
+	}
+	for key, quantity := range newResource.Limits {
+		if !v.CanResourcesResizeInPlace(string(key)) {
+			continue
+		}
+		container.Resources.Limits[key] = quantity
+	}
+	for key, quantity := range newResource.Requests {
+		if !v.CanResourcesResizeInPlace(string(key)) {
+			continue
+		}
+		container.Resources.Requests[key] = quantity
+	}
+}
+
+// IsContainerUpdateCompleted directly determines whether the current container is vertically updated by the spec and status of the container,
+// which conforms to the k8s community standard
+//
+// lstatus: lastContainerStatus record last resource：last container status => last container spec
+// cspec: container spec
+// cstatus: container status
+//
+//	lstatus   cspec  cstatus
+//
+// 1. exist   exist   exist  => compare cspec and cstatus (important)
+// 2. exist   empty   nil    => change qos is forbidden
+// 3. exist   empty   empty  => change qos is forbidden
+// 4. exist   empty   exist  => change qos is forbidden
+// 5. empty   empty   nil    => this container is not inplace-updated by kruise, ignore
+// 6. empty   exist   exist  => this container is not inplace-updated by kruise, ignore
+// 7. empty   exist   exist  => this container is not inplace-updated by kruise, ignore
+func (v *VerticalUpdate) IsContainerUpdateCompleted(pod *v1.Pod, container *v1.Container, containerStatus *v1.ContainerStatus, lastContainerStatus appspub.InPlaceUpdateContainerStatus) bool {
+	// case 5-7
+	if lastContainerStatus.Resources.Limits == nil && lastContainerStatus.Resources.Requests == nil {
```

Comment:
can we simply check pod.status.resize, if the field is not empty, the update is not completed

Reply:
When we only resize limits，filed `.status.resize` will not be changed by api-server and kubelet